### PR TITLE
pass labelSelector to server side opaquely

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
From @smarterclayton 
> The server is responsible for handling label selection for the most part. There is some level of client side processing possible, but for the most part `label selector` should be able to be passed opaquely.

xref #50140

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
/assign @smarterclayton @liggitt 

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by dixudx** [comment]
ping @smarterclayton @liggitt this PR is ready for review. PTAL. Thanks.

File: pkg/kubectl/cmd/top_node.go [hunk 1]
```diff
@@ -132,23 +132,14 @@ func (o *TopNodeOptions) Validate() error {
...
-			return err
-		}
-	}
 	return nil
 }
 
 func (o TopNodeOptions) RunTopNode() error {
 	var err error
-	selector := labels.Everything()
+	selector := labels.Everything().String()
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/top_node.go]: Make this be the empty string

Comment (smarterclayton) [reply pkg/kubectl/cmd/top_node.go]: Actually, you can leave it as is.

**Comment by smarterclayton** [comment]
Actually, before this merges can you verify that the server side returns an error and that the client interprets it correctly when the user inputs an invalid selector (does not parse) for `get` / another command?

I want to ensure that the experience remains the same or roughly compatible when receiving an error, even if we're moving to the server.

**Comment by dixudx** [comment]
> verify that the server side returns an error and that the client interprets it correctly when the user inputs an invalid selector (does not parse) for get / another command?

@smarterclayton It works as expected when user inputs an selector (does not parse, simply passing the raw string).

1. For **valid** label selector

```
$ go run cmd/kubectl/kubectl.go get pods --namespace=kube-system -l component=kube-apiserver,tier=control-plane --v=8
...
...
I1006 05:16:24.420359   19789 round_trippers.go:414] GET https://172.31.6.73:6443/api/v1/namespaces/kube-system/pods?labelSelector=component%3Dkube-apiserver%2Ctier%3Dcontrol-plane
I1006 05:16:24.420411   19789 round_trippers.go:421] Request Headers:
I1006 05:16:24.420492   19789 round_trippers.go:424]     Accept: application/json
I1006 05:16:24.420515   19789 round_trippers.go:424]     User-Agent: kubectl/v0.0.0 (linux/amd64) kubernetes/$Format
I1006 05:16:24.429529   19789 round_trippers.go:439] Response Status: 200 OK in 8 milliseconds
I1006 05:16:24.429601   19789 round_trippers.go:442] Response Headers:
I1006 05:16:24.429627   19789 round_trippers.go:445]     Content-Length: 4027
I1006 05:16:24.429649   19789 round_trippers.go:445]     Date: Fri, 06 Oct 2017 05:16:24 GMT
I1006 05:16:24.429669   19789 round_trippers.go:445]     Content-Type: application/json
I1006 05:16:24.429725   19789 request.go:836] Response Body: {"kind":"PodList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/kube-system/pods","resourceVersion":"2560"},"items":[{"metadata":{"name":"kube-apiserver-ip-172-31-6-73","namespace":"kube-system","selfLink":"/api/v1/namespaces/kube-system/pods/kube-apiserver-ip-172-31-6-73","uid":"e4c72bb4-aa4e-11e7-a328-06b42f1be676","resourceVersion":"1672","creationTimestamp":"2017-10-06T04:29:08Z","labels":{"component":"kube-apiserver","tier":"control-plane"},"annotations":{"kubernetes.io/config.hash":"b46bdf82a93e6042f42454e44523f413","kubernetes.io/config.mirror":"b46bdf82a93e6042f42454e44523f413","kubernetes.io/config.seen":"2017-10-06T04:27:35.989040729Z","kubernetes.io/config.source":"file","scheduler.alpha.kubernetes.io/critical-pod":""}},"spec":{"volumes":[{"name":"k8s-certs","hostPath":{"path":"/etc/kubernetes/pki","type":"DirectoryOrCreate"}},{"name":"ca-certs","hostPath":{"path":"/etc/ssl/certs","type":"DirectoryOrCreate"}}],"containers":[{"name":"kube-apiserver","image":"gcr.io/google_containers/kube-apiserver-amd64:v1.8.0","command":["kube-apiserver","--requestheader-extra-headers-prefix=X-Remote-Extra-","--tls-private-key-file=/etc/kubernetes/pki/apiserver.key","--secure-port=6443","--enable-bootstrap-token-auth=true","--allow-privileged=true","--service-account-key-file=/etc/kubernetes/pki/sa.pub","--kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key","--requestheader-allowed-names=front-proxy-client","--advertise-address=172.31.6.73","--client-ca-file=/etc/kubernetes/pki/ca.crt","--tls-cert-file=/etc/kubernetes/pki/apiserver.crt","--proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt","--kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname","--requestheader-username-headers=X-Remote-User","--requestheader-group-headers=X-Remote-Group","--proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key","--kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt","--requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt","--insecure-port=0","--admission-control=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,ResourceQuota","--service-cluster-ip-range=10.96.0.0/12","--authorization-mode=Node,RBAC","--etcd-servers=http://127.0.0.1:2379"],"resources":{"requests":{"cpu":"250m"}},"volumeMounts":[{"name":"k8s-certs","readOnly":true,"mountPath":"/etc/kubernetes/pki"},{"name":"ca-certs","readOnly":true,"mountPath":"/etc/ssl/certs"}],"livenessProbe":{"httpGet":{"path":"/healthz","port":6443,"host":"127.0.0.1","scheme":"HTTPS"},"initialDelaySeconds":15,"timeoutSeconds":15,"periodSeconds":10,"successThreshold":1,"failureThreshold":8},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","nodeName":"ip-172-31-6-73","hostNetwork":true,"securityContext":{},"schedulerName":"default-scheduler","tolerations":[{"operator":"Exists","effect":"NoExecute"}]},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-10-06T04:27:41Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-10-06T04:27:50Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-10-06T04:27:41Z"}],"hostIP":"172.31.6.73","podIP":"172.31.6.73","startTime":"2017-10-06T04:27:41Z","containerStatuses":[{"name":"kube-apiserver","state":{"running":{"startedAt":"2017-10-06T04:27:50Z"}},"lastState":{},"ready":true,"restartCount":0,"image":"gcr.io/google_containers/kube-apiserver-amd64:v1.8.0","imageID":"docker-pullable://gcr.io/google_containers/kube-apiserver-amd64@sha256:e7377096f0b88b0fcc5dce1c56aed002f999f095a30676c68b8f686a6bb1e943","containerID":"docker://6fc8d4e5f8f9c0186c216c4355abc474ec057ad75bf372575b2c8b6cbd3a186f"}],"qosClass":"Burstable"}}]}
...
NAME                            READY     STATUS    RESTARTS   AGE
kube-apiserver-ip-172-31-6-73   1/1       Running   0          47m
```

2. For **invalid** labelselector

```
$ go run cmd/kubectl/kubectl.go get pods --namespace=kube-system -l component+kube-apiserver,tier --v=8
...
...
I1006 05:37:41.578276   22641 round_trippers.go:414] GET https://172.31.6.73:6443/api/v1/namespaces/kube-system/pods?labelSelector=component%2Bkube-apiserver%2Ctier
I1006 05:37:41.578284   22641 round_trippers.go:421] Request Headers:
I1006 05:37:41.578289   22641 round_trippers.go:424]     Accept: application/json
I1006 05:37:41.578294   22641 round_trippers.go:424]     User-Agent: kubectl/v0.0.0 (linux/amd64) kubernetes/$Format
I1006 05:37:41.580379   22641 round_trippers.go:439] Response Status: 400 Bad Request in 2 milliseconds
I1006 05:37:41.580390   22641 round_trippers.go:442] Response Headers:
I1006 05:37:41.580394   22641 round_trippers.go:445]     Date: Fri, 06 Oct 2017 05:37:41 GMT
I1006 05:37:41.580398   22641 round_trippers.go:445]     Content-Type: application/json
I1006 05:37:41.580401   22641 round_trippers.go:445]     Content-Length: 432
I1006 05:37:41.580583   22641 request.go:836] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"unable to parse requirement: invalid label key \"component+kube-apiserver\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')","reason":"BadRequest","code":400}
I1006 05:37:41.580820   22641 helpers.go:207] server response object: [{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unable to find \"pods\" that match the selector \"component+kube-apiserver,tier\": unable to parse requirement: invalid label key \"component+kube-apiserver\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')",
  "reason": "BadRequest",
  "code": 400
}]
F1006 05:37:41.580850   22641 helpers.go:120] Error from server (BadRequest): Unable to find "pods" that match the selector "component+kube-apiserver,tier": unable to parse requirement: invalid label key "component+kube-apiserver": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
exit status 255
```

> GET https://172.31.6.73:6443/api/v1/namespaces/kube-system/pods?labelSelector=component%2Bkube-apiserver%2Ctier

From above, we can see the raw string has been sent to apiserver. And the error is just what we expected.

@smarterclayton I think it's ready for getting merged.

**Pull request merged**